blog/content/post/json-2.md

```diff
@@ -0,0 +1,202 @@
+---
+date: "2016-07-23T14:46:01+07:00"
+draft: false
+title: "JSON decoding numbers"
+description: "Decode arbitrary JSON data contain number"
+tags: [ "golang", "json", "encode", "decode", "interface", "reflection"]
+categories:
+  - "Standard Library"
+  - "Reflection"
+---
+
+### Preface
+
+At previous post, we had encountered a problem when decode arbitray JSON data containing number. To recap, we decode JSON data into a `map[string]interface{}` and suffered type change, from `int` to `float64`, see more at [JSON Pitfalls]({{< relref "post/json-1.md" >}}).
+
+
+This issue was mentioned at another post too, please see [50 Mistakes](http://devs.cloudimmunity.com/gotchas-and-common-mistakes-in-go-golang/index.html#json_num)
+
+That blog author proposed these options:
+
+1. Using the float value as-is
+2. Convert the float value to the integer type you need.
+3. Using a Decoder type to unmarshal JSON using `json.Number`.
+4. Using a `struct` type that maps your numeric value to the numeric type you need.
+5. Using a `struct` type that maps your number value to `json.RawMessage`
+
+In my opiion, the best way to handling that should always the **option 4** which to not decode arbitrary data at all. We should know struct we are decoding to. There's other benefits of that beside handling number. We can use JSON tag to vaidate input data, so there's no more overhead to check type, check nil, empty value after that. [json.Marshal](https://golang.org/pkg/encoding/json/#Marshal)
```

Comment:
best way to handle that is the ...

 We can use JSON tag to validate

 empty value afterward